diff --git a/data/language-specs/default-language-specs/yaml.lang b/data/language-specs/default-language-specs/yaml.lang
index 99db2ca..cc65138 100644
--- a/data/language-specs/default-language-specs/yaml.lang
+++ b/data/language-specs/default-language-specs/yaml.lang
@@ -21,14 +21,17 @@
 <language id="yaml" name="YAML" _section="Other" version="2.0">
   <metadata>
     <property name="mimetypes">application/x-yaml</property>
-    <property name="globs">*.yaml;*.yml;</property>
+    <property name="globs">*.yaml;*.yml</property>
     <property name="line-comment-start">#</property>
+    <property name="suggested-suffix">.yml</property>
   </metadata>
 
   <styles>
     <style id="scalar"    name="Block literal"   map-to="def:string"/>
     <style id="comment"   name="Comment"         map-to="def:comment"/>
     <style id="string"    name="String"          map-to="def:string"/>
+    <style id="escape-sequence"
+                          name="Escape sequence" map-to="def:special-char"/>
     <style id="anchor"    name="Anchor"          map-to="def:identifier"/>
     <style id="tag"       name="Tag"             map-to="def:preprocessor"/>
     <style id="alias"     name="Alias"           map-to="def:type"/>
@@ -48,8 +51,15 @@
     <define-regex id="rschar">[\s\]\},]</define-regex>
 
     <context id="scalar" class-disabled="no-spell-check">
-      <start>(?&lt;=^|\s)[|&gt;]$</start>
+      <start extended="true">
+        (?&lt;= ^ | \s)
+        [|&gt;] # block scalar indicator
+        (?: [-+]?[0-9]? | [0-9]?[-+]?) # optional indentation and chomping indicators
+        (?: \s+ (\#.*)?)? # optional whitespace and comment
+        $
+      </start>
       <include>
+        <context sub-pattern="1" where="start" style-ref="comment"/>
         <context end-parent="true" style-ref="scalar">
           <start>^(?'indent'\s+)</start>
           <end>^(?!\%{indent@start})</end>
@@ -72,9 +82,32 @@
       <match>(?&lt;=^|\s)&amp;\%{ischar}+(?=$|\s)</match>
     </context>
 
-    <context id="string" end-at-line-end="true" style-ref="string" class-disabled="no-spell-check">
-      <start>(?&lt;=^|\%{lschar})(?'q'["'])</start>
-      <end>[^\\]\%{q@start}</end>
+    <context id="double-quoted-string" style-ref="string" class-disabled="no-spell-check">
+      <start>(?&lt;=^|\%{lschar})"</start>
+      <end>"</end>
+      <include>
+        <context style-ref="escape-sequence">
+          <match extended="true">
+            \\ (?:
+              [0abt\tnvfre "/\\N_LP] |
+              x [0-9a-fA-F]{2} |
+              u [0-9a-fA-F]{4} |
+              U [0-9a-fA-F]{8} |
+              $
+            )
+          </match>
+        </context>
+      </include>
+    </context>
+
+    <context id="single-quoted-string" style-ref="string" class-disabled="no-spell-check">
+      <start>(?&lt;=^|\%{lschar})'</start>
+      <end>'</end>
+      <include>
+        <context style-ref="escape-sequence">
+          <match>''</match>
+        </context>
+      </include>
     </context>
 
     <context id="unquoted-string" end-at-line-end="true" style-ref="string" class-disabled="no-spell-check">
@@ -133,7 +166,8 @@
         <context ref="alias"/>
         <context ref="tag"/>
         <context ref="anchor"/>
-        <context ref="string"/>
+        <context ref="single-quoted-string"/>
+        <context ref="double-quoted-string"/>
         <context ref="inline-null"/>
         <context ref="inline-bool"/>
         <context ref="inline-int"/>
@@ -155,7 +189,8 @@
         <context ref="alias"/>
         <context ref="tag"/>
         <context ref="anchor"/>
-        <context ref="string"/>
+        <context ref="single-quoted-string"/>
+        <context ref="double-quoted-string"/>
         <context ref="inline-null"/>
         <context ref="inline-bool"/>
         <context ref="inline-int"/>
@@ -167,8 +202,22 @@
       </include>
     </context>
 
+
     <context id="map">
+      <!--
+        The following can cause a stack-overflow in libpcre if the map entry
+        contains lots of HTTP urls. If you have a better design for this, please
+        contribute an improvement. Another option might be for GRegex to be
+        compiled without recurse support, but that doesn't fix things that
+        are out in the wild already.
+
+        To generate the failure case try:
+
+        python3 -c 'print("key:", "http://example.com " * 1000)' > testcase.yml
+
       <match>(?&lt;=^|\s)((?:[^:]|:(?=\S))+)(?::\s+|:$)</match>
+      -->
+      <match>(?&lt;=^|\s)([^:\s][^:]*)(?::\s+|:$)</match>
       <include>
         <context sub-pattern="1" style-ref="map-key"/>
       </include>
@@ -202,8 +251,9 @@
         <context ref="alias"/>
         <context ref="tag"/>
         <context ref="anchor"/>
+        <context ref="single-quoted-string"/>
+        <context ref="double-quoted-string"/>
         <context ref="map"/>
-        <context ref="string"/>
         <context ref="unquoted-string"/>
       </include>
     </context>
