diff --git a/data/language-specs/default-language-specs/html.lang b/data/language-specs/default-language-specs/html.lang
index 5db6baa..30145c2 100644
--- a/data/language-specs/default-language-specs/html.lang
+++ b/data/language-specs/default-language-specs/html.lang
@@ -6,6 +6,7 @@
  Copyright (C) 2005-2007 - Marco Barisione <barisione@gmail.com>
  Copyright (C) 2005-2007 - Emanuele Aina
  Copyright (C) 2014 - Sébastien Wilmet <swilmet@gnome.org>
+ Copyright (C) 2018 - Jeffery To <jeffery.to@gmail.com>
 
  GtkSourceView is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
@@ -42,6 +43,7 @@
     <style id="h3" name="Heading level 3" map-to="def:heading3"/>
     <style id="h4" name="Heading level 4" map-to="def:heading4"/>
     <style id="h5" name="Heading level 5" map-to="def:heading5"/>
+    <style id="h6" name="Heading level 6" map-to="def:heading6"/>
   </styles>
 
   <default-regex-options case-sensitive="false"/>
@@ -50,7 +52,7 @@
     <!-- Html comments are more permissive than xml comments -->
     <context id="comment" style-ref="comment" class="comment">
       <start>&lt;!--</start>
-      <end>--\s*&gt;</end>
+      <end>--&gt;</end>
       <include>
         <context ref="def:in-comment"/>
       </include>
@@ -70,100 +72,177 @@
       <end></end>
     </context>
 
+    <context id="heading-tag" once-only="true">
+      <start>\%{def:always-match}</start>
+      <end>&gt;</end>
+      <include>
+        <context sub-pattern="0" where="end" style-ref="tag"/>
+        <context ref="generic-tag"/>
+      </include>
+    </context>
+
+    <define-regex id="h1-close">&lt;/h1\s*&gt;</define-regex>
+    <define-regex id="h2-close">&lt;/h2\s*&gt;</define-regex>
+    <define-regex id="h3-close">&lt;/h3\s*&gt;</define-regex>
+    <define-regex id="h4-close">&lt;/h4\s*&gt;</define-regex>
+    <define-regex id="h5-close">&lt;/h5\s*&gt;</define-regex>
+    <define-regex id="h6-close">&lt;/h6\s*&gt;</define-regex>
+
     <context id="headings">
       <include>
-        <context id="title" style-inside="true" style-ref="title" end-at-line-end="true">
-          <start>&lt;title&gt;</start>
-          <end>&lt;/title&gt;</end>
-          <include>
-            <context sub-pattern="0" where="start" style-ref="tag"/>
-            <context sub-pattern="0" where="end" style-ref="tag"/>
-            <context ref="embedded-lang-hook"/>
-          </include>
-        </context>
-        <context id="h1" end-at-line-end="true">
-          <start>&lt;\s*h1</start>
+        <context id="title" class="no-spell-check">
+          <start>&lt;title\%]</start>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
             <context ref="generic-tag"/>
-            <context id="h1-inside" style-inside="true" style-ref="h1" end-at-line-end="true" end-parent="true">
+
+            <context style-inside="true" style-ref="title" end-parent="true" class-disabled="no-spell-check">
               <start>&gt;</start>
-              <end>&lt;\s*/h1\s*&gt;</end>
+              <end>&lt;/title\s*&gt;</end>
               <include>
                 <context sub-pattern="0" where="start" style-ref="tag"/>
                 <context sub-pattern="0" where="end" style-ref="tag"/>
                 <context ref="embedded-lang-hook"/>
-                <!-- We cannot match any further here, because matching to tag will override our end-match -->
               </include>
             </context>
           </include>
         </context>
-        <context id="h2" end-at-line-end="true">
-          <start>&lt;\s*h2</start>
+
+        <context id="h1" class="no-spell-check">
+          <start>&lt;h1\%]</start>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
-            <context ref="generic-tag"/>
-            <context id="h2-inside" style-inside="true" style-ref="h2" end-at-line-end="true" end-parent="true">
-              <start>&gt;</start>
-              <end>&lt;\s*/h2\s*&gt;</end>
+            <context ref="heading-tag"/>
+
+            <context style-ref="h1" class-disabled="no-spell-check">
+              <start>(?!\%{h1-close})</start>
               <include>
-                <context sub-pattern="0" where="start" style-ref="tag"/>
-                <context sub-pattern="0" where="end" style-ref="tag"/>
+                <context end-parent="true">
+                  <match>(?=\%{h1-close})</match>
+                </context>
                 <context ref="embedded-lang-hook"/>
-                <!-- We cannot match any further here, because matching to tag will override our end-match -->
+                <context ref="tag"/>
               </include>
             </context>
+
+            <context style-ref="tag" end-parent="true">
+              <match>\%{h1-close}</match>
+            </context>
           </include>
         </context>
-        <context id="h3" end-at-line-end="true">
-          <start>&lt;\s*h3</start>
+
+        <context id="h2" class="no-spell-check">
+          <start>&lt;h2\%]</start>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
-            <context ref="generic-tag"/>
-            <context id="h3-inside" style-inside="true" style-ref="h3" end-at-line-end="true" end-parent="true">
-              <start>&gt;</start>
-              <end>&lt;\s*/h3\s*&gt;</end>
+            <context ref="heading-tag"/>
+
+            <context style-ref="h2" class-disabled="no-spell-check">
+              <start>(?!\%{h2-close})</start>
               <include>
-                <context sub-pattern="0" where="start" style-ref="tag"/>
-                <context sub-pattern="0" where="end" style-ref="tag"/>
+                <context end-parent="true">
+                  <match>(?=\%{h2-close})</match>
+                </context>
                 <context ref="embedded-lang-hook"/>
-                <!-- We cannot match any further here, because matching to tag will override our end-match -->
+                <context ref="tag"/>
               </include>
             </context>
+
+            <context style-ref="tag" end-parent="true">
+              <match>\%{h2-close}</match>
+            </context>
           </include>
         </context>
-        <context id="h4" end-at-line-end="true">
-          <start>&lt;\s*h4</start>
+
+        <context id="h3" class="no-spell-check">
+          <start>&lt;h3\%]</start>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
-            <context ref="generic-tag"/>
-            <context id="h4-inside" style-inside="true" style-ref="h4" end-at-line-end="true" end-parent="true">
-              <start>&gt;</start>
-              <end>&lt;\s*/h4\s*&gt;</end>
+            <context ref="heading-tag"/>
+
+            <context style-ref="h3" class-disabled="no-spell-check">
+              <start>(?!\%{h3-close})</start>
               <include>
-                <context sub-pattern="0" where="start" style-ref="tag"/>
-                <context sub-pattern="0" where="end" style-ref="tag"/>
+                <context end-parent="true">
+                  <match>(?=\%{h3-close})</match>
+                </context>
                 <context ref="embedded-lang-hook"/>
-                <!-- We cannot match any further here, because matching to tag will override our end-match -->
+                <context ref="tag"/>
               </include>
             </context>
+
+            <context style-ref="tag" end-parent="true">
+              <match>\%{h3-close}</match>
+            </context>
           </include>
         </context>
-        <context id="h5" end-at-line-end="true">
-          <start>&lt;\s*h5</start>
+
+        <context id="h4" class="no-spell-check">
+          <start>&lt;h4\%]</start>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
-            <context ref="generic-tag"/>
-            <context id="h5-inside" style-inside="true" style-ref="h5" end-at-line-end="true" end-parent="true">
-              <start>&gt;</start>
-              <end>&lt;\s*/h5\s*&gt;</end>
+            <context ref="heading-tag"/>
+
+            <context style-ref="h4" class-disabled="no-spell-check">
+              <start>(?!\%{h4-close})</start>
               <include>
-                <context sub-pattern="0" where="start" style-ref="tag"/>
-                <context sub-pattern="0" where="end" style-ref="tag"/>
+                <context end-parent="true">
+                  <match>(?=\%{h4-close})</match>
+                </context>
                 <context ref="embedded-lang-hook"/>
-                <!-- We cannot match any further here, because matching to tag will override our end-match -->
+                <context ref="tag"/>
               </include>
             </context>
+
+            <context style-ref="tag" end-parent="true">
+              <match>\%{h4-close}</match>
+            </context>
+          </include>
+        </context>
+
+        <context id="h5" class="no-spell-check">
+          <start>&lt;h5\%]</start>
+          <include>
+            <context sub-pattern="0" where="start" style-ref="tag"/>
+            <context ref="heading-tag"/>
+
+            <context style-ref="h5" class-disabled="no-spell-check">
+              <start>(?!\%{h5-close})</start>
+              <include>
+                <context end-parent="true">
+                  <match>(?=\%{h5-close})</match>
+                </context>
+                <context ref="embedded-lang-hook"/>
+                <context ref="tag"/>
+              </include>
+            </context>
+
+            <context style-ref="tag" end-parent="true">
+              <match>\%{h5-close}</match>
+            </context>
+          </include>
+        </context>
+
+        <context id="h6" class="no-spell-check">
+          <start>&lt;h6\%]</start>
+          <include>
+            <context sub-pattern="0" where="start" style-ref="tag"/>
+            <context ref="heading-tag"/>
+
+            <context style-ref="h6" class-disabled="no-spell-check">
+              <start>(?!\%{h6-close})</start>
+              <include>
+                <context end-parent="true">
+                  <match>(?=\%{h6-close})</match>
+                </context>
+                <context ref="embedded-lang-hook"/>
+                <context ref="tag"/>
+              </include>
+            </context>
+
+            <context style-ref="tag" end-parent="true">
+              <match>\%{h6-close}</match>
+            </context>
           </include>
         </context>
       </include>
@@ -213,18 +292,18 @@
     </context>
 
     <context id="style" class="no-spell-check">
-      <start>&lt;\s*style\%]</start>
+      <start>&lt;style\%]</start>
       <include>
         <context sub-pattern="0" where="start" style-ref="tag"/>
         <context ref="generic-tag"/>
 
         <context end-parent="true" style-ref="tag">
-          <match>/\s*&gt;</match>
+          <match>/&gt;</match>
         </context>
 
         <context id="style-code" end-parent="true">
           <start>&gt;</start>
-          <end>&lt;\s*/\s*style\s*&gt;</end>
+          <end>&lt;/style\s*&gt;</end>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
             <context sub-pattern="0" where="end" style-ref="tag"/>
@@ -236,18 +315,18 @@
     </context>
 
     <context id="script" class="no-spell-check">
-      <start>&lt;\s*script\%]</start>
+      <start>&lt;script\%]</start>
       <include>
         <context sub-pattern="0" where="start" style-ref="tag"/>
         <context ref="generic-tag"/>
 
         <context end-parent="true" style-ref="tag">
-          <match>/\s*&gt;</match>
+          <match>/&gt;</match>
         </context>
 
         <context id="js-code" end-parent="true">
           <start>&gt;</start>
-          <end>&lt;\s*/\s*script\s*&gt;</end>
+          <end>&lt;/script\s*&gt;</end>
           <include>
             <context sub-pattern="0" where="start" style-ref="tag"/>
             <context sub-pattern="0" where="end" style-ref="tag"/>
@@ -259,8 +338,8 @@
     </context>
 
     <context id="tag" class="no-spell-check">
-      <start>&lt;\s*/?\s*[a-z0-9_-]+</start>
-      <end>/?\s*&gt;</end>
+      <start>&lt;/?[a-z0-9_-]+</start>
+      <end>/?&gt;</end>
       <include>
         <context sub-pattern="0" where="start" style-ref="tag"/>
         <context sub-pattern="0" where="end" style-ref="tag"/>
diff --git a/data/language-specs/default-language-specs/markdown.lang b/data/language-specs/default-language-specs/markdown.lang
index 2ee8fbf..9a46d36 100644
--- a/data/language-specs/default-language-specs/markdown.lang
+++ b/data/language-specs/default-language-specs/markdown.lang
@@ -160,6 +160,11 @@
       <match>(?&lt;!`)``[^`]+((`|`{3,})[^`]+)*``(?!`)</match>
     </context>
 
+    <context id="3-backticks-code-span" class="no-spell-check" style-ref="code">
+      <start>^```.*$</start>
+      <end>^```$</end>
+    </context>
+
     <!-- Example:
          > Quoted text.
          > Quoted text with `code span`.
@@ -175,6 +180,7 @@
         <context sub-pattern="1" where="start" style-ref="blockquote-marker"/>
         <context ref="1-backtick-code-span"/>
         <context ref="2-backticks-code-span"/>
+        <context ref="3-backticks-code-span"/>
         <context ref="automatic-link"/>
         <context ref="inline-link"/>
         <context ref="reference-link"/>
@@ -371,6 +377,7 @@
         <context ref="code-block"/>
         <context ref="1-backtick-code-span"/>
         <context ref="2-backticks-code-span"/>
+        <context ref="3-backticks-code-span"/>
         <context ref="blockquote"/>
         <context ref="automatic-link"/>
         <context ref="inline-link"/>
diff --git a/data/language-specs/default-language-specs/yaml.lang b/data/language-specs/default-language-specs/yaml.lang
index 99db2ca..cc65138 100644
--- a/data/language-specs/default-language-specs/yaml.lang
+++ b/data/language-specs/default-language-specs/yaml.lang
@@ -21,14 +21,17 @@
 <language id="yaml" name="YAML" _section="Other" version="2.0">
   <metadata>
     <property name="mimetypes">application/x-yaml</property>
-    <property name="globs">*.yaml;*.yml;</property>
+    <property name="globs">*.yaml;*.yml</property>
     <property name="line-comment-start">#</property>
+    <property name="suggested-suffix">.yml</property>
   </metadata>
 
   <styles>
     <style id="scalar"    name="Block literal"   map-to="def:string"/>
     <style id="comment"   name="Comment"         map-to="def:comment"/>
     <style id="string"    name="String"          map-to="def:string"/>
+    <style id="escape-sequence"
+                          name="Escape sequence" map-to="def:special-char"/>
     <style id="anchor"    name="Anchor"          map-to="def:identifier"/>
     <style id="tag"       name="Tag"             map-to="def:preprocessor"/>
     <style id="alias"     name="Alias"           map-to="def:type"/>
@@ -48,8 +51,15 @@
     <define-regex id="rschar">[\s\]\},]</define-regex>
 
     <context id="scalar" class-disabled="no-spell-check">
-      <start>(?&lt;=^|\s)[|&gt;]$</start>
+      <start extended="true">
+        (?&lt;= ^ | \s)
+        [|&gt;] # block scalar indicator
+        (?: [-+]?[0-9]? | [0-9]?[-+]?) # optional indentation and chomping indicators
+        (?: \s+ (\#.*)?)? # optional whitespace and comment
+        $
+      </start>
       <include>
+        <context sub-pattern="1" where="start" style-ref="comment"/>
         <context end-parent="true" style-ref="scalar">
           <start>^(?'indent'\s+)</start>
           <end>^(?!\%{indent@start})</end>
@@ -72,9 +82,32 @@
       <match>(?&lt;=^|\s)&amp;\%{ischar}+(?=$|\s)</match>
     </context>
 
-    <context id="string" end-at-line-end="true" style-ref="string" class-disabled="no-spell-check">
-      <start>(?&lt;=^|\%{lschar})(?'q'["'])</start>
-      <end>[^\\]\%{q@start}</end>
+    <context id="double-quoted-string" style-ref="string" class-disabled="no-spell-check">
+      <start>(?&lt;=^|\%{lschar})"</start>
+      <end>"</end>
+      <include>
+        <context style-ref="escape-sequence">
+          <match extended="true">
+            \\ (?:
+              [0abt\tnvfre "/\\N_LP] |
+              x [0-9a-fA-F]{2} |
+              u [0-9a-fA-F]{4} |
+              U [0-9a-fA-F]{8} |
+              $
+            )
+          </match>
+        </context>
+      </include>
+    </context>
+
+    <context id="single-quoted-string" style-ref="string" class-disabled="no-spell-check">
+      <start>(?&lt;=^|\%{lschar})'</start>
+      <end>'</end>
+      <include>
+        <context style-ref="escape-sequence">
+          <match>''</match>
+        </context>
+      </include>
     </context>
 
     <context id="unquoted-string" end-at-line-end="true" style-ref="string" class-disabled="no-spell-check">
@@ -133,7 +166,8 @@
         <context ref="alias"/>
         <context ref="tag"/>
         <context ref="anchor"/>
-        <context ref="string"/>
+        <context ref="single-quoted-string"/>
+        <context ref="double-quoted-string"/>
         <context ref="inline-null"/>
         <context ref="inline-bool"/>
         <context ref="inline-int"/>
@@ -155,7 +189,8 @@
         <context ref="alias"/>
         <context ref="tag"/>
         <context ref="anchor"/>
-        <context ref="string"/>
+        <context ref="single-quoted-string"/>
+        <context ref="double-quoted-string"/>
         <context ref="inline-null"/>
         <context ref="inline-bool"/>
         <context ref="inline-int"/>
@@ -167,8 +202,22 @@
       </include>
     </context>
 
+
     <context id="map">
+      <!--
+        The following can cause a stack-overflow in libpcre if the map entry
+        contains lots of HTTP urls. If you have a better design for this, please
+        contribute an improvement. Another option might be for GRegex to be
+        compiled without recurse support, but that doesn't fix things that
+        are out in the wild already.
+
+        To generate the failure case try:
+
+        python3 -c 'print("key:", "http://example.com " * 1000)' > testcase.yml
+
       <match>(?&lt;=^|\s)((?:[^:]|:(?=\S))+)(?::\s+|:$)</match>
+      -->
+      <match>(?&lt;=^|\s)([^:\s][^:]*)(?::\s+|:$)</match>
       <include>
         <context sub-pattern="1" style-ref="map-key"/>
       </include>
@@ -202,8 +251,9 @@
         <context ref="alias"/>
         <context ref="tag"/>
         <context ref="anchor"/>
+        <context ref="single-quoted-string"/>
+        <context ref="double-quoted-string"/>
         <context ref="map"/>
-        <context ref="string"/>
         <context ref="unquoted-string"/>
       </include>
     </context>
