diff --git a/src/fonts-pangocairo.cc b/src/fonts-pangocairo.cc
index ea261424c43c5dffc5b7ca56631e6589ec59e3ee..048700135b65b30bbab7bf7f865e57530e10b3b8 100644
--- a/src/fonts-pangocairo.cc
+++ b/src/fonts-pangocairo.cc
@@ -234,7 +234,7 @@ FontInfo::FontInfo(vte::glib::RefPtr<PangoContext> context)
 
         measure_font();
 
-#if PANGO_VERSION_CHECK(1, 44, 0)
+#if 0
         /* Try using the font's metrics; see issue#163. */
         if (auto metrics = vte::take_freeable
             (pango_context_get_metrics(context.get(),
